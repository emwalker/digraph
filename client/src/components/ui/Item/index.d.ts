import React, { ReactNode } from 'react';
import { Color } from 'components/types';
type Topic = {
    displayName: string;
    id: string;
};
type Props = {
    canEdit: boolean;
    children: ReactNode;
    className: string;
    description?: string | null;
    formIsOpen: boolean;
    newlyAdded: boolean;
    repoColors: Color[];
    showEditButton: boolean;
    showLink?: boolean;
    showRepoOwnership: boolean;
    title: string;
    toggleForm: () => void;
    topics: Topic[];
    url: string | null;
};
export default function Item(props: Props): React.JSX.Element;
export {};
