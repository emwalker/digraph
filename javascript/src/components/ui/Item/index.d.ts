import { Component, ReactNode } from 'react';
import { LocationType } from 'components/types';
declare type Topic = {
    displayName: string;
    resourcePath: string;
} | null;
declare type Props = {
    canEdit: boolean;
    children: ReactNode;
    className: string;
    description?: string | null;
    displayColor: string | null;
    formIsOpen: boolean;
    newlyAdded: boolean;
    showEditButton: boolean | null;
    showLink?: boolean;
    orgLogin: string;
    repoName: string | null;
    title: string;
    toggleForm: () => void;
    topics: Topic[];
    url: string | null;
};
declare class Item extends Component<Props> {
    static defaultProps: {
        description: null;
        showLink: boolean;
    };
    get className(): string;
    get showEditButton(): boolean;
    get style(): {
        borderLeft?: undefined;
    } | {
        borderLeft: string;
    };
    get url(): JSX.Element | null;
    get titleLink(): JSX.Element;
    locationDescriptor: (pathname: string, itemTitle: string) => LocationType;
    renderTopicBadge: (topic: Topic) => JSX.Element | null;
    renderEditable: () => JSX.Element;
    renderWide: () => JSX.Element;
    render: () => JSX.Element;
}
export default Item;
