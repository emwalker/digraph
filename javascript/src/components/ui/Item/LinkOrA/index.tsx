import React, { ReactNode } from 'react'
import { Link } from 'found'

import isExternal from 'utils/isExternal'
import { LocationType } from 'components/types'

type Props = {
  children: ReactNode,
  className: string,
  to: LocationType,
}

const LinkOrA = ({ children, className, to }: Props) => {
  if (isExternal(to.pathname)) return <a href={to.pathname} className={className}>{ children }</a>
  return <Link to={to} className={className}>{ children }</Link>
}

export default LinkOrA
