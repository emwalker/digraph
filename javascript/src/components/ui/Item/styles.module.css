.disabledLink {
  color: #959da5;
}
