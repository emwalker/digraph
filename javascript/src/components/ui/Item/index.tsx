import React, { ReactNode } from 'react'
import classNames from 'classnames'

import { Color } from 'components/types'
import { topicPath } from 'components/helpers'
import LinkOrA from './LinkOrA'
import TopicBadge from '../TopicBadge'
import RepoOwnership from '../RepoOwnership'

/* eslint no-underscore-dangle: 0 */

type Topic = {
  displayName: string,
  id: string,
}

function locationDescriptor(pathname: string, itemTitle: string) {
  return {
    pathname,
    query: {},
    search: '',
    state: {
      itemTitle,
    },
  }
}

function CloseButton({ toggleForm }: { toggleForm: () => void }) {
  return (
    <dl className="form-group">
      <button className="btn-link" onClick={toggleForm} type="button">Close</button>
    </dl>
  )
}

function Url({ url, showLink }: { url: string | null, showLink: boolean }) {
  if (!url || !showLink) return null

  return (
    <div
      className="mt-1 link-url branch-name css-truncate css-truncate-target"
    >
      {url}
    </div>
  )
}

function TitleLink({ url, title }: { url: string | null, title: string }) {
  if (!url) {
    return (
      <a
        className="Box-row-link"
        href="#"
      >
        {title}
      </a>
    )
  }

  const to = locationDescriptor(url, title)

  return (
    <LinkOrA to={to} className="Box-row-link">
      {title}
    </LinkOrA>
  )
}

function OuterTopicBadge({ topic }: { topic: Topic }) {
  const { id, displayName } = topic

  return (
    <TopicBadge
      key={id}
      displayName={displayName}
      to={locationDescriptor(topicPath(id), displayName)}
    />
  )
}

type WideItemProps = {
  description: string | null,
  showLink: boolean,
  title: string,
  topics: Topic[],
  url: string | null,
}

function ItemView({ url, description, topics, title, showLink }: WideItemProps) {
  return (
    <div className="clearfix d-flex flex-items-center">
      <div className="col-12">
        <div>
          <TitleLink url={url} title={title} />
          <div>{description}</div>
        </div>
        <Url url={url} showLink={showLink} />
        <div>
          {topics.map((topic) => <OuterTopicBadge key={topic.id} topic={topic} />)}
        </div>
      </div>
    </div>
  )
}

type EditableItemProps = {
  children: ReactNode,
  description: string | null,
  formIsOpen: boolean,
  showEditButton: boolean,
  showLink: boolean,
  title: string,
  toggleForm: () => void,
  topics: Topic[],
  url: string | null,
}

function ItemEditForm({
  formIsOpen, description, url, topics, showEditButton, toggleForm, title, children, showLink,
}: EditableItemProps) {
  return (
    <div>
      <div className="clearfix d-flex flex-items-center">
        <div className="col-10">
          <div>
            <TitleLink url={url} title={title} />
            <div>{description}</div>
          </div>
          <Url url={url} showLink={showLink} />
          <div>
            {topics.map((topic) => <OuterTopicBadge key={topic.id} topic={topic} />)}
          </div>
        </div>
        <div className="col-2 text-center">
          {showEditButton
            ? (
              <button
                className="btn-link"
                onClick={toggleForm}
                type="button"
              >
                Edit
              </button>
            )
            : <a className="color-fg-muted no-underline">Edit</a>}
        </div>
      </div>
      <div>
        {formIsOpen && children}
      </div>
    </div>
  )
}

type Props = {
  canEdit: boolean,
  children: ReactNode,
  className: string,
  description?: string | null,
  formIsOpen: boolean,
  newlyAdded: boolean,
  repoColors: Color[],
  showEditButton: boolean,
  showLink?: boolean,
  showRepoOwnership: boolean,
  title: string,
  toggleForm: () => void,
  topics: Topic[],
  url: string | null,
}

export default function Item(props: Props) {
  const className = classNames('Item-row Box-row', props.className,
    { 'anim-fade-in': props.newlyAdded })
  const showEditButton = !props.formIsOpen && props.showEditButton === true

  const inner = props.canEdit
    ? (
      <ItemEditForm
        children={props.children}
        description={props.description || null}
        formIsOpen={props.formIsOpen}
        showEditButton={showEditButton}
        showLink={props.showLink || false}
        title={props.title}
        toggleForm={props.toggleForm}
        topics={props.topics}
        url={props.url}
      />
    )
    : (
      <ItemView
        description={props.description || null}
        showLink={props.showLink || false}
        title={props.title}
        topics={props.topics}
        url={props.url}
      />
    )

  return (
    <li className={className} key={props.url}>
      {inner}

      <div className="d-flex mb-1 mt-2 flex-items-center">
        <div className="mr-auto">
          <RepoOwnership
            showRepoOwnership={props.showRepoOwnership}
            repoColors={props.repoColors}
          />
        </div>

        {!showEditButton && <CloseButton toggleForm={props.toggleForm} />}
      </div>
    </li>
  )
}
