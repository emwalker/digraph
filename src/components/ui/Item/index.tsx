import React, { Component, ReactNode } from 'react'
import classNames from 'classnames'

import { LocationType } from 'components/types'
import LinkOrA from './LinkOrA'
import TopicBadge from '../TopicBadge'
import styles from './styles.module.css'

/* eslint no-underscore-dangle: 0 */

type Topic = {
  displayName: string,
  resourcePath: string,
} | null

type Props = {
  canEdit: boolean,
  children: ReactNode,
  className: string,
  description?: string | null,
  displayColor: string | null,
  formIsOpen: boolean,
  newlyAdded: boolean,
  showEditButton: boolean | null,
  showLink?: boolean,
  orgLogin: string,
  repoName: string | null,
  title: string,
  toggleForm: () => void,
  topics: Topic[],
  url: string | null,
}

class Item extends Component<Props> {
  static defaultProps = {
    description: null,
    showLink: false,
  }

  get className(): string {
    return classNames(
      'Item-row',
      'Box-row',
      this.props.className,
      { 'anim-fade-in': this.props.newlyAdded },
    )
  }

  get showEditButton(): boolean {
    return !this.props.formIsOpen && this.props.showEditButton === true
  }

  get style() {
    if (!this.props.displayColor) return {}

    return {
      borderLeft: `5px solid ${this.props.displayColor}`,
    }
  }

  get url() {
    if (!this.props.url || !this.props.showLink) return null

    return (
      <div
        className="mt-1 link-url branch-name css-truncate css-truncate-target"
      >
        { this.props.url }
      </div>
    )
  }

  get titleLink() {
    if (!this.props.url) {
      return (
        <a
          className="Box-row-link"
          href="#"
        >
          {this.props.title}
        </a>
      )
    }

    const to = this.locationDescriptor(this.props.url, this.props.title)

    return (
      <LinkOrA to={to} className="Box-row-link">
        { this.props.title }
      </LinkOrA>
    )
  }

  locationDescriptor = (pathname: string, itemTitle: string): LocationType => (
    {
      pathname,
      query: {},
      search: '',
      state: {
        orgLogin: this.props.orgLogin,
        repoName: this.props.repoName,
        itemTitle,
      },
    }
  )

  renderTopicBadge = (topic: Topic) => {
    if (!topic) return null
    const { resourcePath, displayName } = topic
    return (
      <TopicBadge
        key={resourcePath}
        displayName={displayName}
        to={this.locationDescriptor(resourcePath, displayName)}
      />
    )
  }

  renderEditable = () => (
    <>
      <div className="clearfix d-flex flex-items-center">
        <div className="col-10">
          <div>
            { this.titleLink }
            <div>{ this.props.description }</div>
          </div>
          { this.url }
          <div>
            { this.props.topics.map(this.renderTopicBadge) }
          </div>
        </div>
        <div className="col-2 text-center">
          { this.showEditButton
            ? (
              <button
                className="btn-link"
                onClick={this.props.toggleForm}
                type="button"
              >
                Edit
              </button>
            )
            : <span className={styles.disabledLink}>Edit</span>}
        </div>
      </div>
      <div>
        { this.props.formIsOpen && this.props.children }
      </div>
    </>
  )

  renderWide = () => (
    <div className="clearfix d-flex flex-items-center">
      <div className="col-12">
        <div>
          { this.titleLink }
          <div>{ this.props.description }</div>
        </div>
        { this.url }
        <div>
          { this.props.topics.map(this.renderTopicBadge) }
        </div>
      </div>
    </div>
  )

  render = () => {
    const { url } = this.props

    return (
      <li
        className={this.className}
        style={this.style}
        key={url}
      >
        { this.props.canEdit
          ? this.renderEditable()
          : this.renderWide()}
      </li>
    )
  }
}

export default Item
