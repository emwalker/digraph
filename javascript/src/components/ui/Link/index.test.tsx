import React from 'react'
import { shallow } from 'enzyme'

import { UnwrappedLink as Link } from './index'
import Item from '../Item'

const props = {
  link: {
    title: 'title',
    url: 'url',
  },
  orgLogin: 'gnusto',
  relay: {},
  view: {
    currentRepository: {
      name: 'General collection',
    },
  },
  viewer: {},
}

describe('<Link />', () => {
  // @ts-expect-error
  const wrapper = shallow(<Link {...props} />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('passes the showLink flag to the Item component', () => {
    const item = wrapper.find(Item)
    expect(item.prop('showLink')).toBeTruthy()
  })
})
