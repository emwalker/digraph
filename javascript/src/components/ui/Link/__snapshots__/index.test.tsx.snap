// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> renders 1`] = `
<Item
  canEdit={true}
  className="Box-row--link"
  description={null}
  displayColor="transparent"
  formIsOpen={false}
  orgLogin="gnusto"
  repoName="General collection"
  showEditButton={true}
  showLink={true}
  title="title"
  toggleForm={[Function]}
  topics={Array []}
  url="url"
>
  <EditLinkContainer
    isOpen={false}
    link={
      Object {
        "title": "title",
        "url": "url",
      }
    }
    orgLogin="gnusto"
    relay={Object {}}
    toggleForm={[Function]}
  />
</Item>
`;
